src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -72,4 +92,223 @@ object RuleUtils {
       None // logicalPlan is non-linear or it has no LogicalRelation.
     }
   }
+
+  /**
+   * Get replacement plan for current plan. The replacement plan reads data from indexes.
+   * If HybridScan is enabled, additional logical plans for the appended data would be
+   * generated and merged with index data plan. Refer [[getComplementIndexPlan]].
+   *
+   * Pre-requisites
+   * - We know for sure the index which can be used to replace the plan.
+   *
+   * NOTE: This method currently only supports replacement of Scan Nodes i.e. Logical relations
+   *
+   * @param spark Spark session
+   * @param index index used in replacement plan
+   * @param plan current plan
+   * @return replacement plan
+   */
+  def getReplacementPlan(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      plan: LogicalPlan,
+      useBucketSpec: Boolean): LogicalPlan = {
+    // Here we can't replace the plan completely with the index. This will create problems.
+    // For e.g. if Project(A,B) -> Filter(C = 10) -> Scan (A,B,C,D,E)
+    // if we replace this plan with index Scan (A,B,C), we lose the Filter(C=10) and it will
+    // lead to wrong results. So we only replace the base relation.
+    val hybridScanEnabled = spark.sessionState.conf
+      .getConfString(
+        IndexConstants.INDEX_HYBRID_SCAN_ENABLED,
+        IndexConstants.INDEX_HYBRID_SCAN_ENABLED_DEFAULT.toString)
+      .toBoolean
+
+    if (hybridScanEnabled) {
+      getHybridScanIndexPlan(spark, index, plan, useBucketSpec)
+    } else {
+      getIndexPlan(spark, index, plan, useBucketSpec)
+    }
+  }
+
+  private def getIndexPlan(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      plan: LogicalPlan,
+      useBucketSpec: Boolean): LogicalPlan = {
+    plan transformDown {
+      case baseRelation @ LogicalRelation(_: HadoopFsRelation, baseOutput, _, _) =>
+        val location =
+          new InMemoryFileIndex(spark, Seq(new Path(index.content.root)), Map(), None)
+        val relation = HadoopFsRelation(
+          location,
+          new StructType(),
+          StructType(index.schema.filter(baseRelation.schema.contains(_))),
+          if (useBucketSpec) Some(index.bucketSpec) else None,
+          new ParquetFileFormat,
+          Map())(spark)
+
+        val updatedOutput =
+          baseOutput.filter(attr => relation.schema.fieldNames.contains(attr.name))
+        baseRelation.copy(relation = relation, output = updatedOutput)
+    }
+  }
+
+  private def getHybridScanIndexPlan(
```

Comment:
nit: Can you comment on all methods and classes, where applicable to make everything consistent? For instance, `getHybridScanIndexPlan` does not have any documentation but `getComplementIndexPlan` has documentation.